# Zero to Hero

## 1.06 Hackeando Kaggle

Voy a **hackear** Kaggle  para descubrir cuantos  "BAJA+2"  hay en el mes 202101, que es donde  *no* tengo la clase

La idea es subir a Kaggle una prediccion que tenga TODOS 1,  que se le envíe estímulo a todos
<br>
Es obvio que esa prediccion va a dar una pésima ganancia
<br>
Pero me va a permitir tener un sistema de dos ecuaciones con dos incognitas, y despejando tendre la cantidad de BAJA+2 que hay en los datos del futuro

In [2]:
rm(list = ls())
gc()

library("data.table") # cargo la libreria  data.table
library("rpart") # cargo la libreria  rpart
library("rpart.plot")

options(repr.plot.width = 20, repr.plot.height = 10)
setwd("~/buckets/b1/")


used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  654155 35.0 1446000    77.3 1374135  73.4
Vcells 1217625  9.3 8388608    64.0 1924981  14.7

Leo el dataset

In [3]:
dataset <- fread("~/datasets/vivencial_dataset_pequeno.csv") # cargo el dataset


Cuento la clase

In [4]:
dataset[, .N, list(foto_mes, clase_ternaria)]


foto_mes clase_ternaria N     
1 202107   CONTINUA       162211
2 202107   BAJA+1           1137
3 202107   BAJA+2           1248
4 202109                  165093

Defino  el data.table  dfuturo que va a tener solo los datos de  202109

In [5]:
dfuturo <- dataset[foto_mes == 202109]


cuento la cantidad de lineas del dataset

In [6]:
nrow(dfuturo)


[1] 165093

Creo un vector que sean los  ids

In [7]:
vector_ids <- dfuturo[, numero_de_cliente]


In [8]:
head(vector_ids)


[1] 29183733 29184468 29185245 29186441 29186475 29187730

Creo un vector de todos unos, con la cantidad de registros que tiene dfuturo
<br>
utilizo la instruccion  **rep()**  de R  https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/rep

In [9]:
vector_enviar <- rep(1, nrow(dfuturo))


In [10]:
head(vector_enviar)


[1] 1 1 1 1 1 1

In [11]:
length(vector_enviar)


[1] 165093

Excelente, tiene tantos unos como registros tiene dfuturo

creo una tabla con  lo que voy a enviar a Kaggle

In [12]:
tabla_final <- as.data.table(list(
    "numero_de_cliente" = vector_ids,
    "Predicted" = vector_enviar
))


In [13]:
head(tabla_final)


numero_de_cliente Predicted
1 29183733          1        
2 29184468          1        
3 29185245          1        
4 29186441          1        
5 29186475          1        
6 29187730          1

In [14]:
# genero el archivo para Kaggle
# creo la carpeta donde va el experimento
dir.create("./exp/", showWarnings = FALSE)
dir.create("./exp/ZH2016/", showWarnings = FALSE)

# genero el archivo para Kaggle
fwrite(tabla_final,
        file = "./exp/ZH2016/todos_unos.csv",
        sep = ","
)


Subo el archivo a Kaggle y me fijo cuanta ganancia me da en el Public Leaderboard

da una ganancia de **-357.447**   ( menos 357 millones )

### el hackeo

Ahora pasamos a calcular la cantidad de BAJA+2  que hay en los datos de septiembre-2021,  hackeamos Kaggle

Tengo dos incognitas, por un lado lo que llamo POS , los positivos, las "BAJA+2"
<br>
Por otro lado, lo que llamo NEG, los negativos,  los "BAJA+1" y "CONTINUA"

### Se que se cumplen estas dos cosas:
* POS + NEG = 165237
* 117000\*POS - 3000\*NEG =-357447000

Es un sistema de dos ecuaciones con dos incógnitas, has multiples formas de resolverlo, la primera sería despejar POS.
<br>
pero utilizamos esta página  https://www.wolframalpha.com/calculators/system-equation-calculator

## Al correrlo obtenemos que  POS = 1152, o sea **las BAJA+2  de 202109  son  1152**

## ¿Qué nos dicen estas 1152 BAJA+2 de septiembre del 2021? ¿Explican algo?
<br>
Las BAJA+2 que hay en julio del 2021 son  1268
<br>
1248 > 1152
<br>
Con lo cual es de esperar mayores ganancias de los modelos predictivos en 202107  que en 202109,  va a irnos PEOR en el Public Leaderboard que en nuestros traning/testing.